src/client/QXmppOmemoManager.cpp

```diff
@@ -0,0 +1,5349 @@
+void QXmppOmemoManager::setClient(QXmppClient *client)
+{
+    QXmppClientExtension::setClient(client);
+    client->setEncryptionExtension(this);
+
+    auto trustManager = client->findExtension<QXmppTrustManager>();
+    if (!trustManager) {
+        qFatal("QXmppTrustManager object is not available, it must be added to client before adding QXmppOmemoManager object to client");
+    }
+    d->trustManager = trustManager;
+
+    auto pubSubManager = client->findExtension<QXmppPubSubManager>();
+    if (!pubSubManager) {
+        qFatal("QXmppPubSubManager object is not available, it must be added to client before adding QXmppOmemoManager object to client");
+    }
+    d->pubSubManager = pubSubManager;
+
+    connect(trustManager, &QXmppTrustManager::trustLevelsChanged, this, [=](const QHash<QString, QMultiHash<QString, QByteArray>> &modifiedKeys) {
+        const auto modifiedOmemoKeys = modifiedKeys.value(ns_omemo_2);
+        emit trustLevelsChanged(modifiedOmemoKeys);
+
+        for (auto itr = modifiedOmemoKeys.cbegin(); itr != modifiedOmemoKeys.cend(); ++itr) {
+            const auto keyOwnerJid = itr.key();
+            const auto keyId = itr.value();
+
+            // Emit 'deviceChanged()' only if there is a device with the key.
+            const auto devices = d->devices.value(keyOwnerJid);
+            for (auto itr = devices.cbegin(); itr != devices.cend(); ++itr) {
+                if (itr->keyId == keyId) {
+                    emit deviceChanged(keyOwnerJid, itr.key());
+                    return;
+                }
+            }
+        }
+    });
+}
+
+bool QXmppOmemoManager::handlePubSubEvent(const QDomElement &element, const QString &pubSubService, const QString &nodeName)
+{
+    if (nodeName == ns_omemo_2_devices) {
+        if (QXmppPubSubEvent<QXmppOmemoDeviceListItem>::isPubSubEvent(element)) {
+            QXmppPubSubEvent<QXmppOmemoDeviceListItem> event;
+            event.parse(element);
+
+            // TODO: Can the following condition be omitted or is there a use case for it?
+            if (event.eventType() == QXmppPubSubEventBase::Items) {
+                const auto items = event.items();
+
+                // Only process items if the event notification contains one.
+                // That is necessary because PubSub allows publishing without
+                // items leading to notification-only events.
+                if (!items.isEmpty()) {
+                    const auto deviceListItem = items.constFirst();
+                    if (deviceListItem.id() == QXmppPubSubManager::standardItemIdToString(QXmppPubSubManager::Current)) {
+                        d->updateDevices(pubSubService, event.items().constFirst());
+                    } else {
+                        d->handleIrregularDeviceListChanges(pubSubService);
+                    }
+                }
+
+                return true;
+            }
+        } else {
+            QXmppPubSubEvent event;
+            event.parse(element);
+
```

Comment:
Isn't it possible to handle both in one by only using PubSubEvent<QXmppOmemoDeviceListItem>? That was my intention when writing the PubSubEvent class, if it doesn't work we should fix that.